### PR TITLE
[Java] deserialize to LinkedHashSet when unique items

### PR DESCRIPTION
**Pull request by borsch**
fix issue: https://github.com/OpenAPITools/openapi-generator/issues/10167

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by wing328** [comment]
CI reports the following failure. Can you please take a look when you've time?

```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/java/feign/src/main/java/org/openapitools/client/model/Pet.java:[210,40] error: incompatible types: Class<LinkedHashSet> cannot be converted to Class<? extends JsonDeserializer>
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project petstore-feign: Compilation failure
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/java/feign/src/main/java/org/openapitools/client/model/Pet.java:[210,40] error: incompatible types: Class<LinkedHashSet> cannot be converted to Class<? extends JsonDeserializer>
[ERROR] -> [Help 1]
```

Ref: https://app.circleci.com/pipelines/github/OpenAPITools/openapi-generator/12357/workflows/53530b0f-82c0-4a08-8a42-77cc4cf6b5bc/jobs/25986

**Comment by wing328** [comment]
Here is another one:
```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/jaxrs/jersey2/src/gen/java/org/openapitools/model/Pet.java:[184,41] incompatible types: java.lang.Class<java.util.LinkedHashSet> cannot be converted to java.lang.Class<? extends com.fasterxml.jackson.databind.JsonDeserializer>
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project jaxrs-jersey-petstore-server: Compilation failure
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/jaxrs/jersey2/src/gen/java/org/openapitools/model/Pet.java:[184,41] incompatible types: java.lang.Class<java.util.LinkedHashSet> cannot be converted to java.lang.Class<? extends com.fasterxml.jackson.databind.JsonDeserializer>
[ERROR] -> [Help 1]
```

Ref: https://app.circleci.com/pipelines/github/OpenAPITools/openapi-generator/12357/workflows/53530b0f-82c0-4a08-8a42-77cc4cf6b5bc/jobs/25986/parallel-runs/1?filterBy=FAILED => then click on node 1

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 1]
```diff
@@ -1257,6 +1258,8 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
                 model.imports.add("ArrayList");
             } else if ("set".equals(property.containerType)) {
                 model.imports.add("LinkedHashSet");
+                model.imports.add("JsonDeserialize");
+                property.vendorExtensions.put("x-setter-extra-annotation", "@JsonDeserialize(using = LinkedHashSet.class)");
```
Comment (sonalidhome) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:1262]: I have also tried the same in one of the scenarios. But gave me compilation errors. The workaround works as expected when you modify `@JsonDeserialize(using = LinkedHashSet.class)` to `@JsonDeserialize(as = LinkedHashSet.class)`.

Comment (borsch) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java]: thanks @sonalidhome. fixed

**Comment by borsch** [comment]
@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @nmuesch (2021/01) @wing328

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java (hunk 1 above)
Comment (sonalidhome) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java]: There is compilation error.  Can you please try to re-generate and push it? 
`[ERROR] COMPILATION ERROR : 
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/java/retrofit2/src/main/java/org/openapitools/client/model/Pet.java:[18,49] package com.fasterxml.jackson.databind.annotation does not exist
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/java/retrofit2/src/main/java/org/openapitools/client/model/Pet.java:[205,4] cannot find symbol
  symbol:   class JsonDeserialize
  location: class org.openapitools.client.model.Pet
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.1:compile (default-compile) on project petstore-retrofit2: Compilation failure: Compilation failure:
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/java/retrofit2/src/main/java/org/openapitools/client/model/Pet.java:[18,49] package com.fasterxml.jackson.databind.annotation does not exist
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/client/petstore/java/retrofit2/src/main/java/org/openapitools/client/model/Pet.java:[205,4] cannot find symbol
[ERROR] symbol:   class JsonDeserialize
[ERROR] location: class org.openapitools.client.model.Pet
[ERROR] -> [Help 1]`

Comment (borsch) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java]: @sonalidhome thanks. finally should be ok now

**Comment by borsch** [comment]
@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @nmuesch (2021/01) @wing328

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java [hunk 2]
```diff
@@ -1228,6 +1229,8 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
                 model.imports.add("ArrayList");
             } else if ("set".equals(property.containerType)) {
                 model.imports.add("LinkedHashSet");
+                model.imports.add("JsonDeserialize");
+                property.vendorExtensions.put("x-setter-extra-annotation", "@JsonDeserialize(as = LinkedHashSet.class)");
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java:1233]: FYI. I may rename it to `x-java-setter-extra-annotation` later as this extension is for the Java generators only.

**Comment by MelleD** [comment]
@borsch and @wing328 
This PR broke the JsonNullable deserialization. If you use a JsonNullable with uniqueItems. This code will generated:

e.g.
```
@JsonDeserialize(as = LinkedHashSet.class)
public void setFoos(JsonNullable<Set<String>> foos) {
  this.foos = foos;
}
```

Exception

```[.c.j.MappingJackson2HttpMessageConverter] : Failed to evaluate Jackson deserialization for type [[simple type, class FooDto]]: com.fasterxml.jackson.databind.JsonMappingException: Failed to narrow type [simple type, class org.openapitools.jackson.nullable.JsonNullable<java.util.Set<java.lang.String>>] with annotation (value java.util.LinkedHashSet), from 'setFoos': Class `java.util.LinkedHashSet` not subtype of `org.openapitools.jackson.nullable.JsonNullable<java.util.Set<java.lang.String>>`
```

```
foos:
  type: array
  uniqueItems: true
  items:
    type: string
  maxItems: 10
  nullable: true
```

**Pull request merged**